### PR TITLE
Add WebSocket Terminal support

### PR DESCRIPTION
**Pull request by isc-bsaviano**
This PR adds support for for the new terminal endpoint added in Atelier API v7. It will be available in the next preview of IRIS 2023.2. See the new documentation included in this PR for how to start the terminal and what features it supports.

### CODE REVIEW

**Review by isc-rsingh — APPROVED** [review]

File: docs/RunDebug.md [hunk 1]
```diff
@@ -128,3 +128,37 @@ If you are experiencing issues using the debugger, please follow these steps bef
...
+The WebSocket terminal can be opened from [the command palette](https://code.visualstudio.com/docs/getstarted/userinterface#_command-palette) using the `ObjectScript: Launch WebSocket Terminal` command. The WebSocket terminal connection will be established using the current server connection. A WebSocket terminal connection can also be opened from the [Terminal Profiles menu](https://code.visualstudio.com/docs/terminal/basics#_terminal-shells).
+
+## Troubleshooting WebSocket Terminal Issues
+
+If you are experiencing issues using the WebSocket terminal, please follow these steps before opening an issue on GitHub:
+
+1. Open a standard terminal on your server and `zn` to the namespace containing the class or routine you are debugging.
+2. Run the command `Kill ^IRIS.Temp.Atelier("terminal")`, then `Set ^IRIS.Temp.Atelier("terminal") = 1` to turn on the Atelier API terminal logging feature.
+3. In VS Code, launch the WebSocket terminal and run the commands that produce the error.
+4. Once the error appears, copy the contents of the `^IRIS.Temp.Atelier("terminal")` global and add it to your GitHub issue.
```
Comment (gjsjohnmurray) [inline docs/RunDebug.md:163]: If this trace global might contain confidential data then we should probably advise people to mask that before posting.

**Comment by gjsjohnmurray** [comment]
I look forward to trying this out as soon as a suitable 2023.2 preview becomes available.

**Comment by isc-bsaviano** [comment]
This can be tested with the [2nd 2023.2 developer preview kit](https://community.intersystems.com/post/developer-preview-2-intersystems-iris-iris-health-20232) (2023.2.0.201.0)

**Comment by daimor** [comment]
That's very interesting

**Comment by daimor** [comment]
Well, 
The terminal thing itself looks good, looks working

But there are a few Issues

When connection is disable, it keeps showing in this add terminal menu
image

But, in the list of main commands shown only when any ObjectScript file is open, even if the connection is active

And in the project with a Python virtual environment, made with venv, VSCode activates this virtual environment when opened a new terminal. No idea if it can be configured somehow, probably it's just implemented a way, that it sends this command to any terminal.
image

to reproduce it, 
create a new virtual environment in the project and switch to it, VSCode should notice a new environment and offer to switch to it
```
python -m venv venv
source venv/bin/activate
```

File: package.json [hunk 2]
```diff
@@ -343,6 +345,10 @@
         {
           "command": "vscode-objectscript.openErrorLocation",
           "when": "vscode-objectscript.connectActive && workspaceFolderCount != 0"
+        },
+        {
+          "command": "vscode-objectscript.launchWebSocketTerminal",
+          "when": "editorLangId =~ /^objectscript/ && vscode-objectscript.connectActive"
```
Comment (daimor) [inline package.json:351]:
```suggestion
          "when": "vscode-objectscript.connectActive"
```

**Review by daimor — CHANGES_REQUESTED** [review]

**Comment by isc-bsaviano** [comment]
@daimor The `terminal.profiles` contribution point doesn't take a `when` clause, so I can' hide it when there's no active connection. I could open a  VS Code issue but I doubt it would be resolved before this PR is merged. The second image looks like a bug in whatever code sends that command. It should check that the terminal isn't an extension terminal before sending it. I'm not sure where you should report that.

**Comment by daimor** [comment]
well, ok then

**Review by daimor — DISMISSED** [review]

**Comment by isc-bsaviano** [comment]
FYI reported the when clause issue: https://github.com/microsoft/vscode/issues/183354

**Comment by gjsjohnmurray** [comment]
For the venv issue see https://github.com/microsoft/vscode-python/issues/20562

File: src/commands/webSocketTerminal.ts [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+
+function terminalConfigForUri(api: AtelierAPI, extensionUri: vscode.Uri): vscode.ExtensionTerminalOptions | undefined {
+  // Make sure the server connection is active
+  if (!api.active || api.ns == "") {
+    vscode.window.showErrorMessage("WebSocket Terminal requires an active server connection.", "Dismiss");
+    return;
+  }
+  // Make sure the server has the terminal endpoint
+  if (api.config.apiVersion < 7) {
+    vscode.window.showErrorMessage("WebSocket Terminal requires Atelier API version 7 or above.", "Dismiss");
```
Comment (gjsjohnmurray) [inline src/commands/webSocketTerminal.ts:604]: I suggest stating the IRIS version rather than the API version, as the latter is meaningless to most.

**Review by gjsjohnmurray — CHANGES_REQUESTED** [review]
Some comments made. I haven't yet had time to install a 2023.2 that I can actually test the terminal with.

File: src/commands/webSocketTerminal.ts [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+    const workspaceFolders = vscode.workspace.workspaceFolders || [];
+    if (workspaceFolders.length == 0) {
+      vscode.window.showErrorMessage("WebSocket Terminal requires an open workspace.", "Dismiss");
+    } else if (workspaceFolders.length == 1) {
+      // Use the current connection
+      uri = workspaceFolders[0].uri;
+    } else {
+      // Pick from the workspace folders
+      uri = (
+        await vscode.window.showWorkspaceFolderPick({
```
[the remark this answers is not in the document]
Comment (isc-bsaviano) [reply src/commands/webSocketTerminal.ts:657]: I think using the prebuilt QuickPick is fine

**Review by gjsjohnmurray — APPROVED** [review]

**Pull request merged**